src/apiClient.ts

```diff
@@ -6,30 +6,55 @@ export interface ApiClientConfig {
   optionServiceUrl?: string
   analyticsServiceUrl?: string
   userNotificationServiceUrl?: string
-  tokenRefreshServiceUrl?: string
   debug?: boolean
 }
 
-export interface Tokens {
-  accessToken?: string
-  refreshToken?: string
-}
-
 export interface Handlers {
   onAccessTokenUpdate?: (token: string) => void
   onFailedTokenRefresh?: () => void
 }
 
 class ApiClient {
-  static instance: ApiClient
+  private static instance: ApiClient
 
-  configuration: ApiClientConfig = {}
-  tokens: Tokens = {}
-  handlers: Handlers = {}
-  version = "v1"
+  configuration: ApiClientConfig
+  accessToken: {
+    token: string | null
+    expires: Date | null
+  }
+  handlers: Handlers
+  version: string
+  runsInBrowser: boolean
+  refreshToken: () => Promise<{ accessToken: string }>
+  ensureTokenFreshness: () => void
 
   constructor() {
+    if (ApiClient.instance) return ApiClient.instance
+
+    this.configuration = {}
+    this.accessToken = {
+      expires: null,
+      token: null,
+    }
+    this.handlers = {}
+    this.version = "v1"
+    this.runsInBrowser = typeof window !== "undefined"
+    this.refreshToken = async () => {
+      throw new Error(
+        "The refreshToken function has not been set the the apiClient instance, use apiClient.setTokenRefreshHandler to pass a function which handles the token refresh"
+      )
+    }
+    this.ensureTokenFreshness = async () => {
+      console.log("ensure fresshenes!!!", Date.now())
+      console.log("ensure fresshenes!!!", this.accessToken)
+      // TODO: check weather the token is n minutes valid if not refreh it
```

Comment:
todo and logs before merging
```suggestion
      // TODO: check whether the token is n minutes valid if not refresh it
```